### PR TITLE
feat(Core/SmartScripts): SMART_EVENT_FLAG_WHILE_CHARMED

### PR DESCRIPTION
**Pull request by IntelligentQuantum**
## Changes Proposed:
-  
-  

## Issues Addressed:

- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/731

## Tests Performed:

- 
- 

## How to Test the Changes:

1.
2.
3.

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Review by Nyeriah — APPROVED** [review]

**Comment by Nyeriah** [comment]
Closes https://github.com/azerothcore/azerothcore-wotlk/issues/731

File: src/server/game/AI/SmartScripts/SmartAI.cpp [hunk 1]
```diff
@@ -517,6 +529,9 @@ void SmartAI::UpdateAI(uint32 diff)
             mFollowArrivedTimer -= diff;
     }
 
+    if (!IsAIControlled())
+        return;
```
Comment (Winfidonarleyan) [inline src/server/game/AI/SmartScripts/SmartAI.cpp:533]: brackets

File: src/server/game/AI/SmartScripts/SmartAI.cpp [hunk 2]
```diff
@@ -681,10 +692,13 @@ void SmartAI::MoveInLineOfSight(Unit* who)
 
     GetScript()->OnMoveInLineOfSight(who);
 
+    if (!IsAIControlled())
+        return;
```
Comment (Winfidonarleyan) [inline src/server/game/AI/SmartScripts/SmartAI.cpp:696]: brackets

File: src/server/game/AI/SmartScripts/SmartAI.cpp [hunk 3]
```diff
@@ -642,12 +653,12 @@ void SmartAI::EnterEvadeMode()
         return;
     }
 
-    me->RemoveEvadeAuras();
+    if (!_EnterEvadeMode())
+        return;
```
Comment (Winfidonarleyan) [inline src/server/game/AI/SmartScripts/SmartAI.cpp:657]: brackets

File: src/server/game/AI/SmartScripts/SmartAI.cpp [hunk 4]
```diff
@@ -700,13 +714,14 @@ void SmartAI::MoveInLineOfSight(Unit* who)
...
-        return false;
-    return true;
+    return !(me->GetReactState() == REACT_PASSIVE);
 }
 
-bool SmartAI::AssistPlayerInCombat(Unit* who)
+bool SmartAI::AssistPlayerInCombatAgainst(Unit* who)
 {
+    if (!IsAIControlled())
+        return false;
```
Comment (Winfidonarleyan) [inline src/server/game/AI/SmartScripts/SmartAI.cpp:723]: brackets

File: src/server/game/AI/SmartScripts/SmartAI.cpp [hunk 5]
```diff
@@ -779,7 +794,8 @@ void SmartAI::JustReachedHome()
 void SmartAI::EnterCombat(Unit* enemy)
 {
     // Xinef: Interrupt channeled spells
-    me->InterruptSpell(CURRENT_CHANNELED_SPELL, true, true);
+    if (IsAIControlled())
+        me->InterruptSpell(CURRENT_CHANNELED_SPELL, true, true);
```
Comment (Winfidonarleyan) [inline src/server/game/AI/SmartScripts/SmartAI.cpp:798]: brackets

File: src/server/game/AI/SmartScripts/SmartAI.cpp [hunk 6]
```diff
@@ -847,6 +863,9 @@ void SmartAI::DamageTaken(Unit* doneBy, uint32& damage, DamageEffectType damaget
     if (doneBy)
         GetScript()->ProcessEventsFor(SMART_EVENT_DAMAGED, doneBy, damage);
 
+    if (!IsAIControlled()) // don't allow players to use unkillable units
+        return;
```
Comment (Winfidonarleyan) [inline src/server/game/AI/SmartScripts/SmartAI.cpp:867]: brackets

File: src/server/game/AI/SmartScripts/SmartAI.cpp [hunk 7]
```diff
@@ -906,9 +921,29 @@ void SmartAI::InitializeAI()
...
 
-void SmartAI::OnCharmed(bool apply)
+void SmartAI::OnCharmed(bool /* apply */)
 {
-    GetScript()->ProcessEventsFor(SMART_EVENT_CHARMED, nullptr, 0, 0, apply);
+    bool const charmed = me->IsCharmed();
+    if (charmed) // do this before we change charmed state, as charmed state might prevent these things from processing
+    {
+        if (HasEscortState(SMART_ESCORT_ESCORTING | SMART_ESCORT_PAUSED | SMART_ESCORT_RETURNING))
+            EndPath(true);
```
Comment (Winfidonarleyan) [inline src/server/game/AI/SmartScripts/SmartAI.cpp:930]: brackets

File: src/server/game/AI/SmartScripts/SmartAI.cpp [hunk 8]
```diff
@@ -906,9 +921,29 @@ void SmartAI::InitializeAI()
...
+    if (!charmed && !me->IsInEvadeMode())
+    {
+        if (mCanRepeatPath)
+            StartPath(mRun, GetScript()->GetPathId(), true);
+        else
+            me->SetWalk(!mRun);
+
+        if (Unit* charmer = me->GetCharmer())
+            AttackStart(charmer);
+    }
```
Comment (Winfidonarleyan) [inline src/server/game/AI/SmartScripts/SmartAI.cpp:944]: brackets

File: src/server/game/AI/SmartScripts/SmartAI.cpp [hunk 9]
```diff
@@ -1001,6 +1036,10 @@ void SmartAI::SetCombatMove(bool on)
         return;
 
     mCanCombatMove = on;
+
+    if (!IsAIControlled())
+        return;
```
Comment (Winfidonarleyan) [inline src/server/game/AI/SmartScripts/SmartAI.cpp:1041]: brackets

File: src/server/game/AI/SmartScripts/SmartScript.cpp [hunk 10]
```diff
@@ -4025,6 +4025,9 @@ void SmartScript::ProcessEvent(SmartScriptHolder& e, Unit* unit, uint32 var0, ui
     if ((e.event.event_phase_mask && !IsInPhase(e.event.event_phase_mask)) || ((e.event.event_flags & SMART_EVENT_FLAG_NOT_REPEATABLE) && e.runOnce))
         return;
 
+    if (!(e.event.event_flags & SMART_EVENT_FLAG_WHILE_CHARMED) && IsCharmedCreature(me))
+        return;
```
Comment (Winfidonarleyan) [inline src/server/game/AI/SmartScripts/SmartScript.cpp:4029]: brackets

**Comment by Lynethris** [comment]
How test this ?

**Comment by IntelligentQuantum** [comment]
> How test this ?

That's my question xD

Read https://github.com/azerothcore/azerothcore-wotlk/pull/1431

**Comment by Kitzunu** [comment]
WHILE CHARMED. Self explanatory, no? :D

**Comment by Lynethris** [comment]
All seems work. Tested with Generosity Abounds, Krolmir, Hammer of Storms and Battling the Elements quests.

**Pull request merged**